packages/react-reconciler/src/ReactFiberUnwindWork.js

```diff
@@ -135,6 +127,21 @@ function createClassErrorUpdate(
       this.componentDidCatch(error, {
         componentStack: stack !== null ? stack : '',
       });
+      if (__DEV__) {
+        if (typeof getDerivedStateFromError !== 'function') {
+          // If componentDidCatch is the only error boundary method defined,
+          // then it needs to call setState to recover from errors.
+          // If no state update is scheduled then the boundary will swallow the error.
+          const updateQueue = fiber.updateQueue;
+          warningWithoutStack(
+            updateQueue !== null && updateQueue.firstUpdate !== null,
```

Comment:
This check isn't sufficient because `firstUpdate` might be a low priority update. It needs to be sync. You can check `fiber.expirationTime === Sync` instead.